back/engines/commercial/insights/spec/services/insights/category_detection_service_spec.rb

```diff
@@ -0,0 +1,141 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe Insights::CategoryDetectionService do
+  describe '#detect_categories' do
+    subject(:service) { described_class.new(nlp_client) }
+    let(:nlp_client) { instance_double(NLP::Api, 'nlp_client') }
+
+    let(:response) {[
+      {
+        "text" => "kaggle",
+        "frequency" => 11
+      },
+      {
+        "text" => "google",
+        "frequency" => 10
+      }
+    ]}
+
+    context 'no input' do
+      let(:view) { create(:view) }
+      it 'returns [] when there\'s no inputs' do
+        allow(nlp_client).to receive(:project_tag_suggestions).and_return(response)
+        result = service.detect_categories(view)
+
+        expect(result.length).to eq(0)
+      end
+    end
+
+    context 'some input' do
+      let(:ideas) { create_list(:idea, 3) }
+      let(:project) { create(:project, ideas: ideas) }
+      let(:view) { create(:view, scope: project) }
+
+      context 'no previous detected_categories' do
+        it 'returns [] and creates nothing when there\'s no suggestions' do
+          allow(nlp_client).to receive(:project_tag_suggestions).and_return(nil)
+          result = service.detect_categories(view)
+
+          aggregate_failures 'check results' do
+            expect(result.length).to eq(0)
+            expect(Insights::DetectedCategory.where(view: view).length).to eq 0
+          end
+        end
+
+        it 'returns [] and lets detected_categories unchanged when there\'s no suggestions' do
+          allow(nlp_client).to receive(:project_tag_suggestions).and_return([])
+          result = service.detect_categories(view)
+
+          aggregate_failures 'check results' do
+            expect(result.length).to eq(0)
+            expect(Insights::DetectedCategory.where(view: view).length).to eq 0
```

Comment:
It would probably make sense to create a detected category before running `detect_categories` and make sure it's still there.

Reply:
Oh, I see there are some tests for that below. But then I don't understand to goal of this one.

Reply:
I was not too sure what NLP could retrun when the detection failed, so I added a few tests to check it woulnd't crash when NLP answered unexpected things. Not the most useful test, but it's a small one too